python/pyarrow/array.pxi

```diff
@@ -75,13 +97,43 @@ cdef class DataType:
         else:
             raise NotImplementedError(str(self))
 
+    property logical_type:
+
+        def __get__(self):
+            cdef Type type_id = self.type.id()
+            if type_id in _logical_type_map:
+                return _logical_type_map[type_id]
```

Comment:
I think we'd have to refactor this quite a bit to make the type id accessible inside of `pandas_compat.py` (breaking out the type related code into `type.pxi`). I was hoping to get this in without that, though I can certainly do that refactor here.